### ISSUE TITLE
[Request]: Update Containers docs for Infrastructure

### ISSUE THREAD

**Issue by roshan-elastic** [elastic/observability-docs#3745]
### Description

We're going to be updating our container template to (a) use the new template we are using for hosts (b) making 'docker containers' support **all** containers.

Initially, this means we need to:

1. Update the [metrics definitions](https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html)

2. Update the [inventory docs](https://www.elastic.co/guide/en/observability/current/view-infrastructure-metrics.html) to reflect these are 'containers' (not just 'docker containers')

_Old view of containers_
![containers - now](https://github.com/elastic/observability-docs/assets/117740680/22eb4bda-05d6-476b-8fa7-afb02ab282ae)

_What it will start to look like (showing 'hosts' for now)_
![containers - next](https://github.com/elastic/observability-docs/assets/117740680/7ada1679-8b75-47a8-8431-f1cf15f5883e)

**Notes:** 
- See [Docker Container Metric Definitions](https://docs.google.com/document/d/1xrLMnT-Nz5TqkPT5PxNjJNLjrU4qXMsbyTWkmqLhtLw/edit#heading=h.lrmgncaurmh9) and [Kubernetes Container Metric Definitions](https://docs.google.com/document/d/1xrLMnT-Nz5TqkPT5PxNjJNLjrU4qXMsbyTWkmqLhtLw/edit#heading=h.k69qexcnfbv8)
- There will be future releases which start to incorporate 'containers' into other views (e.g. from a host or a service) but that's out of scope for now (see [epic for details](https://github.com/elastic/observability-dev/issues/3470))

### Resources

- [x] https://github.com/elastic/observability-dev/issues/3313

### Which documentation set does this change impact?

Stateful and Serverless

### Feature differences

None

### What release is this request related to?

TBC  (likely 8.15)

### Collaboration model

I'm happy to help!

### Point of contact.

**Main contact:** @roshan-elastic  + @jennypavlova + @MiriamAparicio 

**Stakeholders:**
@smith

**Comment by roshan-elastic** [comment]
@colleenmcginnis FYI

**Comment by bmorelli25** [comment]
Thanks for opening this issue. Marking as blocked and adding to our backlog until we know those Docker metric definitions.

**Comment by roshan-elastic** [comment]
Hey @bmorelli25,

We've completed this issue now (barring a minor bug with the logs tab) but hopefully this is good to start documenting?

I'd imagine you'll need some support untangling this.

Here is a summary:

1. Containers in Inventory
2. Container Views

### Containers in Inventory

There are no changes to the UI behaviour for 'Docker Containers' in Inventory except for clicking on one will open the new 'Container View' fly-out (see _2. Container Views_ below).

**Details**

- In Inventory, all container types will continue to show in the waffle map (they show all container runtimes, e.g. Docker, containerd - anything which emits `container.id`):

<img width="1015" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/642db878-9907-4f13-b8b3-5baf304acda3">

- However, there is a 'bug' which we can't get around for the moment which is that the Inventory UI says **Docker Containers** when we are in fact showing ALL containers (this has been the case for a long time).

- This works fine if you happen to only have 'docker' containers, the UI works OK - including the metrics you see per container:

<img width="992" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/93ea3873-4645-45db-91c4-6ab9a2ce310d">

> **Note** : that the 'metrics' used for the 'Docker Containers' are using [docker specific fields](https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html) which only work with docker containers:

<img width="332" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/bafe4c39-9f3b-4318-a442-9372aa5626a7">

However, if you are looking at non-docker containers, the metrics will all show as '0%' or '-' because the docker fields don't exist:

<img width="958" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/7a08c2b3-709d-46b1-8c05-18d0ed5e0bbe">

We do plan to move towards a single 'Container' entity type which works for any runtime (i.e. Docker and other runtimes) but we can't get there yet

### Containers Views

We have updated the container view to be similar to 'Hosts', i.e. there is a fly-out and an updated container view:

**Details**
<img width="1437" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/57ec3849-d704-4e93-8ea1-be64801b4a98">

**Metrics**

Even when there are no [docker](https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html) metrics, the UI will adapt to show the correct metrics based on the runtime (this was the case in the old view too).

At the moment, it will adapt to either [docker metrics](https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html) or Kubernetes metrics (e.g. `average(kubernetes.container.cpu.usage.limit.pct)`). 

> **Note** : You can test by opening any visualisation in Lens but I'm sure @jennypavlova can help

**Alerts**
This is a new section:

- Any alert which matches the `container.id` should show in the list
- If you are viewing a 'Docker' container then you can create an alert against it:

<img width="806" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/b884373f-f9b7-439a-883e-c0a03d936176">

- If you are viewing any other container runtime, you will not be able to create an alert against it:

<img width="804" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/e5808364-13f2-4d12-b46f-bdf9fd8bdb3b">

> **Note** : This is because the 'Inventory' alert rule for 'Docker Containers' will only evaluate 'Docker' metrics (so it doesn't make sense to show metrics which won't work for the current container).

### Why aren't we just showing 'Containers' instead of 'Docker Containers'?

We want to get there but we need to figure out a way to ensure there is a single 'Container' entity that we can show in the UI and create alerts on.

We could change this today if we wanted but the problem is that the inventory 'Docker Container' entity type with the respective 'metrics' is shared between 'Inventory' and the 'Inventory Alerts':

_Inventory - 'Docker Containers' and their metrics_
<img width="575" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/28b3c0f1-9dc5-4e95-a176-134a6839d2dd">

_Inventory Alerts - 'Docker Containers' and their metrics_
<img width="440" alt="image" src="https://github.com/elastic/observability-docs/assets/117740680/d4a4bc1d-aaf4-47b4-b916-374baebd1693">

However, if we were to 'docker containers' to 'containers' and make changes to the default metrics, this would impact existing inventory rules users may have. So, instead, we're moving cautiously and we're figuring out how to have a single 'container' entity type with normalised metrics which work with any container type.

**Comment by roshan-elastic** [comment]
cc @jennypavlova are you able to assist the docs team at all on this?

I'm thinking mainly about field names/usage for the container views as it adapts to docker vs K8s containers.

**Comment by roshan-elastic** [comment]
(if they reach out that is)

**Comment by jennypavlova** [comment]
> cc @jennypavlova are you able to assist the docs team at all on this?

Sure, please let me know if you have any questions, I am happy to help :)

**Comment by roshan-elastic** [comment]
Hey @bmorelli25,

Just wanted to check if this one will have some assigned?

It's not urgent but basically making container views.

**Comment by bmorelli25** [comment]
Hey @roshan-elastic, I have this on our list, but it'll need to take a backseat to ECO and OTel right now. I don't have an estimated date or writer yet.

**Comment by roshan-elastic** [comment]
No problem @bmorelli25 - not a priority.

**Comment by bmorelli25** [comment]
@roshan-elastic do we know what the metrics definitions are?

> Update the [metrics definitions](https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html)...we don't know what these are yet

**Comment by dedemorton** [comment]
@roshan-elastic To add to Brandon's question, I have a request. Can you quickly update this issue to reflect the current state of this request. For example, I'm assuming that this statement is no longer true:

>This is WIP so we'll let you know when we have our first customer-facing release

I also want to make sure we're targeting the correct release. Currently the issue indicates that we are likely targeting 8.15:

>TBC (likely 8.15)

**Comment by roshan-elastic** [comment]
Hey @bmorelli25 @dedemorton,

> Can you quickly update this issue to reflect the current state of this request

Sure.

> I also want to make sure we're targeting the correct release. Currently the issue indicates that we are likely targeting 8.15:

That's right. This change went out in 8.15.

> @roshan-elastic do we know what the metrics definitions are?

Yes, we have them now.

When a container is using Docker, it should be using these definitions:

[Docker Metric Definitions](https://docs.google.com/document/d/1xrLMnT-Nz5TqkPT5PxNjJNLjrU4qXMsbyTWkmqLhtLw/edit#heading=h.lrmgncaurmh9)

When a container is using ContainerD (i.e. Kubernetes) then it should be using these metrics:

[Kubernetes Container Metric Definitions](https://docs.google.com/document/d/1xrLMnT-Nz5TqkPT5PxNjJNLjrU4qXMsbyTWkmqLhtLw/edit#heading=h.k69qexcnfbv8)

> **Note** The view of a service should adapt based on the available metrics for the container

### Question

Would it make sense to repurpose the https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html page into a more generic container page which as sections for Docker Metrics and Kubernetes Container Metrics?

We have short links in the UI which I can remotely change which are pointing towards Kubernetes or Docker metrics in a page (so you can change the URL without breaking any links in the UI - I can change the short-links to point to anywhere):

e.g.

(where the `#VALUE` would refer to the `id=VALUE` in the doc)

https://ela.st/docs-infra-k8s-metrics#key-metrics-memory 
https://ela.st/docs-infra-k8s-metrics#key-metrics-cpu

https://ela.st/docs-infra-docker-container-metrics#key-metrics-cpu
https://ela.st/docs-infra-docker-container-metrics#key-metrics-memory

What I'm thinking is that if we had a single 'container' page with sections for Docker and K8s that I could point towards then https://ela.st/docs-infra-k8s-metrics#key-metrics-memory  could point towards `YOURCHOSENURL#ID_OF_SECTION`?

### More information for the future

**TL;DR** : We're going to normalise all container types in the future to use the same fields for metrics (e.g. `container.cpu.usage` instead of specific ones per runtime like `kubernetes.container.memory.usage.limit.pct` for Kubernetes and `docker.cpu.total.pct`). It's worth bearing in mind that when we make this change, we'll need another docs update to reflect this.

The UI is pretty inconsistent with containers because we're conditionally changing the metric formulae in the container view based on the available metrics. For example, the fields used to show 'CPU' are different between Docker containers and K8s containers.

At the same time, when you go to Inventory, it shows 'Docker Containers' in the drop-down  but it will show **all** containers with a `container.id` (inc containerd ones for K8s).

This is described in more detail in [this epic](https://github.com/elastic/observability-dev/issues/3470) where we're going to eventually normalise all containers as "Containers" (instead of having references to "Docker Containers" with specific docker metrics) and use the generic container fields like `container.cpu.usage`.

We'll make all of these changes so that there are just 'Containers' which use generic fields for metrics and then we'll probably leave the existing metrics (which are used in Inventory and Inventory alert rules) as `legacy Docker CPU` etc.

**Comment by dedemorton** [comment]
>Would it make sense to repurpose the https://www.elastic.co/guide/en/observability/current/docker-container-metrics.html page into a more generic container page which as sections for Docker Metrics and Kubernetes Container Metrics?

@roshan-elastic Sure we can have a single page that contains sections for both docker and K8s metrics.
